src/exodus/wallet.cpp

```diff
@@ -77,88 +90,132 @@ SigmaSpend Wallet::CreateSigmaSpend(PropertyId property, SigmaDenomination denom
     }
 
     // Create spend.
-    SigmaProof proof(DefaultSigmaParams, mint->key, anonimitySet.begin(), anonimitySet.end());
+    auto key = GetKey(mint.get());
+    SigmaProof proof(DefaultSigmaParams, key, anonimitySet.begin(), anonimitySet.end());
 
     if (!VerifySigmaSpend(mint->property, mint->denomination, mint->chainState.group, anonimitySet.size(), proof)) {
         throw WalletError(_("Failed to create spendable spend"));
     }
 
-    return SigmaSpend(SigmaMintId(*mint, DefaultSigmaParams), mint->chainState.group, anonimitySet.size(), proof);
+    return SigmaSpend(SigmaMintId(mint->property, mint->denomination, SigmaPublicKey(key, DefaultSigmaParams)),
+        mint->chainState.group, anonimitySet.size(), proof);
 }
 
 void Wallet::EraseSigmaMint(const SigmaMintId &id)
 {
-    if (!CWalletDB(walletFile).EraseExodusMint(id)) {
-        throw std::runtime_error("Fail to erase dangling mint from wallet");
-    }
+    mintWallet.PushFrontToMintPool(id);
 }
 
 bool Wallet::HasSigmaMint(const SigmaMintId& id)
 {
-    return CWalletDB(walletFile).HasExodusMint(id);
+    return mintWallet.HasMint(id);
+}
+
+bool Wallet::HasSigmaSpend(const secp_primitives::Scalar& serial, SigmaMint &mint)
+{
+    LOCK(pwalletMain->cs_wallet);
+    return mintWallet.HasMint(serial);
 }
 
 SigmaMint Wallet::GetSigmaMint(const SigmaMintId& id)
 {
     SigmaMint mint;
 
-    if (!CWalletDB(walletFile).ReadExodusMint(id, mint)) {
+    CWalletDB walletdb(walletFile);
+    if (!walletdb.ReadExodusHDMint(id, mint)) {
         throw std::invalid_argument("Mint with specified identifier is not exists");
     }
 
     return mint;
 }
 
-void Wallet::SetSigmaMintUsedTransaction(SigmaMintId const &id, uint256 const &tx)
+boost::optional<SigmaMint>
+    Wallet::GetSpendableSigmaMint(PropertyId property, SigmaDenomination denomination)
 {
-    auto mint = GetSigmaMint(id);
+    // Get all spendable mints.
+    LOCK(pwalletMain->cs_wallet);
+    std::vector<SigmaMint> spendables;
+    mintWallet.ListMints(std::back_inserter(spendables), true, true);
 
-    mint.spentTx = tx;
+    auto eraseFrom = std::remove_if(spendables.begin(), spendables.end(), [denomination](SigmaMint const &mint) -> bool {
+        return denomination != mint.denomination;
+    });
+    spendables.erase(eraseFrom, spendables.end());
```

Comment:
We should use `boost::make_function_output_iterator` when invoke `ListMints` instead of erase it like this.